File: src/main/java/minic/ui/workbench/MiniCWorkbenchController.java
package minic.ui;

import minic.uiapi.UiControlResultDto;

import java.util.Objects;

/**
 * 工作台控制动作封装。
 */
public final class MiniCWorkbenchController {
    private final MiniCWorkbenchViewModel viewModel;

    /**
     * 创建工作台控制器。
     *
     * @param viewModel UI 状态模型
     */
    public MiniCWorkbenchController(MiniCWorkbenchViewModel viewModel) {
        this.viewModel = Objects.requireNonNull(viewModel, "viewModel");
    }

    /**
     * 启动默认样例会话。
     */
    public void startDefaultSession() {
        MiniCSampleProgram sample = MiniCSamplePrograms.defaultSample();
        viewModel.loadSource(sample.name(), sample.source());
        viewModel.startSession();
    }

    /**
     * 执行下一步。
     *
     * @return 控制结果
     */
    public UiControlResultDto next() {
        return viewModel.next();
    }

    /**
     * 跳转到下一编译环节。
     *
     * @return 控制结果
     */
    public UiControlResultDto nextStage() {
        return viewModel.nextStage();
    }
}

File: uiweb/src/workbench/MiniCWorkbenchController.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiControlResultDto } from "../translation/uiapi";
import { MiniCSamplePrograms } from "../editor/MiniCSamplePrograms";
import type { MiniCWorkbenchViewModel } from "./MiniCWorkbenchViewModel";

export const miniCWorkbenchControllerMirror = {
  "javaPath": "src/main/java/minic/uilocal/workbench/MiniCWorkbenchController.java",
  "webPath": "uiweb/src/workbench/MiniCWorkbenchController.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCWorkbenchController",
  "kind": "class",
  "imports": [
    "java.util.Objects",
    "minic.uiapi.UiControlResultDto"
  ],
  "fields": [
    {
      "name": "viewModel",
      "signature": "private final MiniCWorkbenchViewModel viewModel"
    }
  ],
  "methods": [
    {
      "name": "next",
      "signature": "next()"
    },
    {
      "name": "nextStage",
      "signature": "nextStage()"
    },
    {
      "name": "startDefaultSession",
      "signature": "startDefaultSession()"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCWorkbenchController {
  static readonly mirror = miniCWorkbenchControllerMirror;

  readonly mirror = miniCWorkbenchControllerMirror;

  constructor(private readonly viewModel: MiniCWorkbenchViewModel) {}

  startDefaultSession(): void {
    const sample = MiniCSamplePrograms.defaultSample();
    void this.viewModel.loadSource(sample.name, sample.source)
      .then(() => this.viewModel.startSession());
  }

  next(): Promise<UiControlResultDto> {
    return this.viewModel.next();
  }

  nextStage(): Promise<UiControlResultDto> {
    return this.viewModel.nextStage();
  }

  summary(): string {
    return `MiniCWorkbenchController: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

export default MiniCWorkbenchController;
